crates/ruff_formatter/src/printer/mod.rs

```diff
@@ -107,7 +107,7 @@ impl<'a> Printer<'a> {
                     self.flush_line_suffixes(queue, stack, Some(element));
                 } else {
                     // Only print a newline if the current line isn't already empty
-                    if self.state.line_width > 0 {
+                    if self.state.line_has_content {
```

Comment:
This was an awkward bug. The form feed character has a width of 0, but we need to print it. I hope this won't hurt performance much (one more field to track and write)